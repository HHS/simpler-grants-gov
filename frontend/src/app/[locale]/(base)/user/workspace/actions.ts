"use server";

import { getSession } from "src/services/auth/session";
import { inviteUserToOrganization } from "src/services/fetch/fetchers/organizationsFetcher";
import { OrganizationInviteRecord } from "src/types/organizationTypes";
import { z } from "zod";

import { getTranslations } from "next-intl/server";

export type OrganizationInviteValidationErrors = {
  email?: string[];
  role?: string[];
};

export type OrganizationInviteResponse = {
  data?: OrganizationInviteRecord;
  errorMessage?: string;
  validationErrors?: OrganizationInviteValidationErrors;
  success?: boolean;
};

const validateInviteUserAction = async (formData: FormData) => {
  const t = await getTranslations("ManageUsers.inviteUser.validationErrors");
  const schema = z.object({
    email: z.string().min(1, { message: t("email") }),
    role: z.string().min(1, { message: t("role") }),
  });

  const validatedFields = schema.safeParse({
    email: formData.get("email"),
    role: formData.get("role"),
  });

  // Return early if the form data is invalid (server side validation!)
  if (!validatedFields.success) {
    return validatedFields.error.flatten().fieldErrors;
  }
};

export const inviteUserAction = async (
  _prevState: unknown,
  formData: FormData,
  organizationId: string,
): Promise<OrganizationInviteResponse> => {
  const session = await getSession();

  if (!session || !session.token || !session.user_id) {
    return {
      errorMessage: "Not logged in",
    };
  }

  const validationErrors = await validateInviteUserAction(formData);
  if (validationErrors) {
    return {
      validationErrors,
    };
  }

  const requestData = {
    email: formData.get("email") as string,
    roleId: formData.get("role") as string,
    organizationId,
  };

  let inviteUserResponse: OrganizationInviteRecord;
  try {
    inviteUserResponse = await inviteUserToOrganization(
      session.token,
      requestData,
    );
    return { data: inviteUserResponse, success: true };
  } catch (e) {
    const error = e as Error;
    console.error(
      `Error inviting user to org - ${error.message} ${error.cause?.toString() || ""}`,
    );
    return {
      errorMessage: error.message,
    };
  }
};
