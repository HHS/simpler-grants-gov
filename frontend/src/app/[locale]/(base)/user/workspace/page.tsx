import { Metadata } from "next";
import { getSession } from "src/services/auth/session";
import withFeatureFlag from "src/services/featureFlags/withFeatureFlag";
import { getUserOrganizations } from "src/services/fetch/fetchers/organizationsFetcher";
import { getUserDetails } from "src/services/fetch/fetchers/userFetcher";
import { Organization } from "src/types/applicationResponseTypes";
import { LocalizedPageProps } from "src/types/intl";

import { useTranslations } from "next-intl";
import { getTranslations } from "next-intl/server";
import Link from "next/link";
import { redirect } from "next/navigation";
import { ErrorMessage, GridContainer } from "@trussworks/react-uswds";

export async function generateMetadata({
  params,
}: LocalizedPageProps): Promise<Metadata> {
  const { locale } = await params;
  const t = await getTranslations({ locale });
  const meta: Metadata = {
    title: t("UserWorkspace.pageTitle"),
    description: t("Index.metaDescription"),
  };
  return meta;
}

const UserOrganizationsList = ({
  userOrganizations,
}: {
  userOrganizations: Organization[];
}) => {
  const t = useTranslations("UserWorkspace");
  return (
    <>
      <h2>{t("organizations")}</h2>
      <ul>
        {userOrganizations.map((userOrganization) => (
          <li key={userOrganization.organization_id}>
            <Link href={`/organization/${userOrganization.organization_id}`}>
              {userOrganization.sam_gov_entity.legal_business_name}
            </Link>
          </li>
        ))}
      </ul>
    </>
  );
};

async function UserWorkspace() {
  const t = await getTranslations("UserWorkspace");

  const session = await getSession();
  if (!session?.email) {
    // this won't happen, as email is required on sessions, and we're wrapping this in an auth gate in the layout
    console.error("no user session, or user has no email address");
    return;
  }
  let userDetails;
  let userOrganizations;
  const userDetailsPromise = getUserDetails(session.token, session.user_id);
  const userOrganizationsPromise = getUserOrganizations(
    session.token,
    session.user_id,
  );
  try {
    [userDetails, userOrganizations] = await Promise.all([
      userDetailsPromise,
      userOrganizationsPromise,
    ]);
  } catch (e) {
    console.error("Unable to fetch user details", e);
  }

  return (
    <GridContainer className="padding-top-2 tablet:padding-y-6">
      <h1>{t("title")}</h1>
      {userDetails && userOrganizations ? (
        <UserOrganizationsList userOrganizations={userOrganizations} />
      ) : (
        <ErrorMessage>{t("fetchError")}</ErrorMessage>
      )}
    </GridContainer>
  );
}

export default withFeatureFlag<object, never>(
  UserWorkspace,
  "userAdminOff",
  () => redirect("/maintenance"),
);
