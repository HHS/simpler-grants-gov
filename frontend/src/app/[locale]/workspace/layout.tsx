import { LayoutProps } from "src/types/generalTypes";

import BetaAlert from "src/components/BetaAlert";
import { AuthenticationGate } from "src/components/user/AuthenticationGate";

export default function WorkspaceLayout({ children }: LayoutProps) {
  return (
    <AuthenticationGate>
      <BetaAlert containerClasses="margin-bottom-5" />
      {children}
    </AuthenticationGate>
  );
}
