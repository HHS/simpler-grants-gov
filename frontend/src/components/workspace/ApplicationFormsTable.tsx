import { ApplicationFormDetail } from "src/types/applicationResponseTypes";
import { CompetitionForms } from "src/types/competitionsResponseTypes";
import { FormDetail } from "src/types/formResponseTypes";

import { useTranslations } from "next-intl";
import Link from "next/link";
import { Table } from "@trussworks/react-uswds";

import { USWDSIcon } from "src/components/USWDSIcon";

export const selectFormsByRequired = ({
  forms,
  required,
}: {
  forms: CompetitionForms;
  required: boolean;
}): FormDetail[] => {
  return forms.reduce<FormDetail[]>((acc, item) => {
    if (item.is_required === required) {
      acc.push(item.form);
    }
    return acc;
  }, []);
};

const selectApplicationForm = ({
  applicationForms,
  formId,
}: {
  applicationForms: ApplicationFormDetail[];
  formId: string;
}) => {
  return applicationForms
    ? applicationForms.find(
        (applicationForm) => applicationForm.form_id === formId,
      )
    : null;
};

export const ApplicationFormsTable = ({
  applicationForms,
  applicationId,
  forms,
}: {
  applicationForms: ApplicationFormDetail[];
  applicationId: string;
  forms: CompetitionForms;
}) => {
  const requiredForms = selectFormsByRequired({ forms, required: true });
  const conditionalRequiredForms = selectFormsByRequired({
    forms,
    required: false,
  });
  const t = useTranslations("Application.competitionFormTable");

  return (
    <>
      <h3>{t("requiredForms")}</h3>
      <ApplicationTable
        forms={requiredForms}
        applicationForms={applicationForms}
        applicationId={applicationId}
      />
      {conditionalRequiredForms.length > 0 && (
        <>
          <h3>{t("conditionalForms")}</h3>
          <p>{t("conditionalFormsDescription")}</p>
          <ApplicationTable
            forms={conditionalRequiredForms}
            applicationForms={applicationForms}
            applicationId={applicationId}
          />
        </>
      )}
    </>
  );
};

const selectApplicationFormById = ({
  forms,
  formId,
}: {
  forms: FormDetail[];
  formId: string;
}): FormDetail | undefined => {
  return forms.find((form) => form.form_id === formId);
};

const ApplicationTable = ({
  applicationForms,
  applicationId,
  forms,
}: {
  applicationForms: ApplicationFormDetail[];
  applicationId: string;
  forms: FormDetail[];
}) => {
  const t = useTranslations("Application.competitionFormTable");

  return (
    <Table className="width-full overflow-wrap">
      <thead>
        <tr>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("status")}
          </th>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("form")}
          </th>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("instructions")}
          </th>
          <th scope="col" className="bg-base-lightest padding-y-205">
            {t("updated")}
          </th>
        </tr>
      </thead>
      <tbody>
        {applicationForms.map((form, index) => (
          <tr key={index}>
            <td data-label={t("status")}>
              <CompetitionStatus
                applicationForms={applicationForms}
                formId={form.form_id}
              />
            </td>
            <td data-label={t("form")}>
              <FormLink
                formId={form.form_id}
                forms={forms}
                applicationId={applicationId}
                appFormId={form.application_form_id}
              />
            </td>
            <td data-label={t("instructions")}>
              <InstructionsLink
                forms={forms}
                formId={form.form_id}
                text={t("downloadInstructions")}
                unavailableText={t("attachmentUnavailable")}
              />
            </td>
            <td data-label={t("updated")}>
              <div> -- </div>
            </td>
          </tr>
        ))}
      </tbody>
    </Table>
  );
};

const CompetitionStatus = ({
  formId,
  applicationForms,
}: {
  applicationForms: ApplicationFormDetail[];
  formId: string;
}) => {
  const t = useTranslations("Application.competitionFormTable.statuses");
  const applicationForm = selectApplicationForm({ formId, applicationForms });

  if (applicationForm?.application_form_status === "in_progress") {
    return (
      <div className="display-flex flex-align-center text-bold icon-active">
        <USWDSIcon name="loop" className="margin-right-2px" />
        {t("in_progress")}
      </div>
    );
  } else if (applicationForm?.application_form_status === "complete") {
    return (
      <div className="display-flex flex-align-center text-bold">
        <USWDSIcon name="check" className="text-primary margin-right-2px" />
        {t("complete")}
      </div>
    );
  } else {
    return <>-</>;
  }
};

const InstructionsLink = ({
  formId,
  forms,
  text,
  unavailableText,
}: {
  formId: string;
  forms: FormDetail[];
  text: string;
  unavailableText: string;
}) => {
  const instructions = selectApplicationFormById({
    forms,
    formId,
  })?.form_instruction;
  const downloadPath = instructions?.download_path || null;
  return (
    <>
      {downloadPath ? (
        <Link
          className="display-flex flex-align-center font-sans-2xs"
          href={downloadPath}
        >
          <USWDSIcon name="save_alt" className="margin-right-05" />
          {text}
        </Link>
      ) : (
        <>{unavailableText}</>
      )}
    </>
  );
};

const FormLink = ({
  formId,
  forms,
  applicationId,
  appFormId,
}: {
  formId: string;
  forms: FormDetail[];
  applicationId: string;
  appFormId: string;
}) => {
  const formName = selectApplicationFormById({
    forms,
    formId,
  })?.form_name;

  return (
    <>
      {formName && (
        <Link
          className="text-bold"
          href={`/workspace/applications/application/${applicationId}/form/${appFormId}`}
        >
          {formName}
        </Link>
      )}
    </>
  );
};
