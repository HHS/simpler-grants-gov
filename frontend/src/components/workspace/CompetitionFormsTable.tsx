import { CompetitionForms } from "src/types/competitionsResponseTypes";
import { FormDetail, FormInstruction } from "src/types/formResponseTypes";

import { useTranslations } from "next-intl";
import Link from "next/link";
import { Table } from "@trussworks/react-uswds";

import { USWDSIcon } from "src/components/USWDSIcon";

export const CompetitionFormsTable = ({
  forms,
  applicationId,
}: {
  forms: CompetitionForms;
  applicationId: string;
}) => {
  const requiredForms = selectForms({ forms, required: true });
  const conditionalRequiredForms = selectForms({ forms, required: false });
  const t = useTranslations("Application.competitionFormTable");

  return (
    <>
      <h3>{t("requiredForms")}</h3>

      <CompetitionTable forms={requiredForms} applicationId={applicationId} />
      <h3>{t("conditionalForms")}</h3>

      <CompetitionTable
        forms={conditionalRequiredForms}
        applicationId={applicationId}
      />
    </>
  );
};

export const selectForms = ({
  forms,
  required,
}: {
  forms: CompetitionForms;
  required: boolean;
}): FormDetail[] => {
  return forms.reduce<FormDetail[]>((acc, item) => {
    if (item.is_required === required) {
      acc.push(item.form);
    }
    return acc;
  }, []);
};

const CompetitionTable = ({
  forms,
  applicationId,
}: {
  forms: FormDetail[];
  applicationId: string;
}) => {
  const t = useTranslations("Application.competitionFormTable");

  return (
    <Table className="width-full overflow-wrap">
      <thead>
        <tr>
          <th scope="col" className="bg-base-lighter">
            {t("status")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("form")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("attachment")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("instructions")}
          </th>
          <th scope="col" className="bg-base-lighter">
            {t("updated")}
          </th>
        </tr>
      </thead>
      <CompetitionTableBody forms={forms} applicationId={applicationId} />
    </Table>
  );
};

const CompetitionTableBody = ({
  forms,
  applicationId,
}: {
  forms: FormDetail[];
  applicationId: string;
}) => {
  const t = useTranslations("Application.competitionFormTable");

  return (
    <tbody>
      {forms.map((form, index) => (
        <tr key={index}>
          <td data-label={t("status")}>--</td>
          <td data-label={t("form")}>
            <Link
              className="text-bold"
              href={`/workspace/applications/application/${applicationId}/form/${form.form_id}`}
            >
              {form.form_name}
            </Link>
          </td>
          <td data-label={t("attachment")}>
            <div> -- </div>
          </td>
          <td data-label={t("instructions")}>
            <InstructionsLink
              instsruction={form.form_instruction}
              text={t("downloadInstructions")}
            />
          </td>
          <td data-label={t("updated")}>
            <div> -- </div>
          </td>
        </tr>
      ))}
    </tbody>
  );
};

const InstructionsLink = ({
  instsruction,
  text,
}: {
  instsruction: FormInstruction;
  text: string;
}) => {
  return (
    <Link
      className="display-flex flex-align-center font-sans-2xs"
      href={instsruction.download_path}
    >
      <USWDSIcon name="save_alt" className="margin-right-05" />
      {text}
    </Link>
  );
};
