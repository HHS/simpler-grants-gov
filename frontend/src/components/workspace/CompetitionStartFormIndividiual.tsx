"use client";

import { useTranslations } from "next-intl";
import { RefObject } from "react";
import {
  Button,
  ErrorMessage,
  FormGroup,
  Label,
  ModalFooter,
  ModalRef,
  ModalToggleButton,
  TextInput,
} from "@trussworks/react-uswds";

const CompetitionStartFormIndividiual = ({
  opportunityTitle,
  error = "",
  loading = false,
  modalRef,
  onChange,
  onClose,
  onSubmit,
  validationError = "",
}: {
  opportunityTitle: string;
  loading?: boolean;
  error?: string;
  modalRef: RefObject<ModalRef | null>;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onClose: () => void;
  onSubmit: () => void;
  validationError?: string;
}) => {
  const t = useTranslations("OpportunityListing");

  return (
    <div className="display-flex flex-align-start">
      <FormGroup error={!!validationError} className="margin-top-1">
        <p className="font-sans-sm" data-testid="opportunity-title">
          {opportunityTitle}
        </p>
        <p className="font-sans-3xs">
          {t("startApplicationModal.requiredText")}
        </p>
        <Label
          id={`label-for-name`}
          key={`label-for-name`}
          htmlFor="application-name"
          className="font-sans-2xs"
        >
          {t("startApplicationModal.name")}
          <span>
            <br /> {t("startApplicationModal.description")}
          </span>
        </Label>
        {validationError && <ErrorMessage>{validationError}</ErrorMessage>}
        {error && <ErrorMessage>{error}</ErrorMessage>}

        <TextInput
          type="text"
          name="application-name"
          id="application-name"
          onChange={onChange}
        />
        <ModalFooter>
          <Button
            onClick={onSubmit}
            type="button"
            data-testid="competition-start-individual-save"
            disabled={!!loading}
          >
            {loading ? "loading..." : t("startApplicationModal.saveButtonText")}
          </Button>
          <ModalToggleButton
            modalRef={modalRef}
            closer
            unstyled
            className="padding-105 text-center"
            onClick={onClose}
          >
            {t("startApplicationModal.cancelButtonText")}
          </ModalToggleButton>
        </ModalFooter>
      </FormGroup>
    </div>
  );
};

export default CompetitionStartFormIndividiual;
