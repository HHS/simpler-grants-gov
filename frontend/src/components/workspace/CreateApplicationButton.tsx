"use client";

import { useUser } from "src/services/auth/useUser";
import { startApplication } from "src/services/fetch/fetchers/clientApplicationFetcher";

import { useRouter } from "next/navigation";
import { useCallback, useState } from "react";
import { Button } from "@trussworks/react-uswds";

const CreateApplicationButton = () => {
  const router = useRouter();
  const { user } = useUser();
  const disabled = !user?.token;

  const [loading, setLoading] = useState<boolean>();

  const handleSubmit = useCallback(() => {
    if (!user?.token) return;
    setLoading(true);
    startApplication(user.token)
      .then((data) => {
        const { applicationId } = data;
        router.push(`/formPrototype/${applicationId}`);
      })
      .catch((error) => {
        console.error(error);
      })
      .finally(() => {
        setLoading(false);
      });
  }, [router, user]);
  return (
    <Button disabled={disabled} onClick={handleSubmit} type="submit">
      {loading ? "loading..." : "Create new application"}
    </Button>
  );
};

export default CreateApplicationButton;
