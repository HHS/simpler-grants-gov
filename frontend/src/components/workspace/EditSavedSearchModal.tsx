"use client";

import clsx from "clsx";
import { useUser } from "src/services/auth/useUser";
import { editSavedSearchName } from "src/services/fetch/fetchers/clientSavedSearchFetcher";

import { useTranslations } from "next-intl";
import { RefObject, useCallback, useMemo, useRef, useState } from "react";
import {
  Button,
  ErrorMessage,
  FormGroup,
  Modal,
  ModalFooter,
  ModalHeading,
  ModalRef,
  ModalToggleButton,
  TextInput,
} from "@trussworks/react-uswds";

import Loading from "src/components/Loading";
import SimplerAlert from "src/components/SimplerAlert";
import { USWDSIcon } from "src/components/USWDSIcon";

function SaveSearchInput({
  validationError,
  updateSavedSearchName,
  id,
}: {
  validationError?: string;
  updateSavedSearchName: (name: string) => void;
  id: string;
}) {
  const t = useTranslations("Search.saveSearch.modal");
  const inputRef = useRef<HTMLInputElement>(null);

  return (
    <FormGroup error={!!validationError}>
      <label htmlFor="saved-search-input">{t("inputLabel")}</label>
      {validationError && <ErrorMessage>{validationError}</ErrorMessage>}
      <div className="usa-search usa-search--big" role="search">
        <TextInput
          ref={inputRef}
          className={clsx("usa-input", "maxw-none", {
            "usa-input--error": !!validationError,
          })}
          id={`edit-saved-search-input-${id}`}
          name={`save-search-${id}`}
          defaultValue={""}
          onChange={(e) => updateSavedSearchName(e.target?.value)}
          type="text"
        />
      </div>
    </FormGroup>
  );
}

function SuccessContent({
  modalRef,
  modalId,
  onClose,
}: {
  modalRef: RefObject<ModalRef | null>;
  modalId: string;
  onClose: () => void;
}) {
  const t = useTranslations("SavedSearches.editModal");
  return (
    <>
      <ModalHeading id={`${modalId}-heading`}>{t("successTitle")}</ModalHeading>
      <ModalFooter>
        <ModalToggleButton
          modalRef={modalRef}
          closer
          unstyled
          className="padding-105 text-center"
          onClick={onClose}
        >
          {t("closeText")}
        </ModalToggleButton>
      </ModalFooter>
    </>
  );
}

// note that this is client component, but not marked as "use client" because to do so
// would cause an error when passing the function prop from parent client component see:
// https://github.com/vercel/next.js/discussions/46795
export function EditSavedSearchModal({
  // onSave,
  savedSearchId,
  editText,
}: {
  // onSave: () => void;
  savedSearchId: string;
  editText: string;
}) {
  const modalId = useMemo(
    () => `edit-save-search-${savedSearchId}`,
    [savedSearchId],
  );

  const t = useTranslations("SavedSearches.editModal");
  const modalRef = useRef<ModalRef>(null);
  const { user } = useUser();

  const [validationError, setValidationError] = useState<string>();
  const [savedSearchName, setSavedSearchName] = useState<string>();
  const [apiError, setApiError] = useState<boolean>();
  const [loading, setLoading] = useState<boolean>();
  const [updated, setUpdated] = useState<boolean>();

  const handleSubmit = useCallback(() => {
    if (validationError) {
      setValidationError(undefined);
    }
    if (!savedSearchName) {
      setValidationError(t("emptyNameError"));
      return;
    }
    setLoading(true);
    editSavedSearchName(savedSearchName, savedSearchId, user?.token)
      .then(() => {
        setUpdated(true);
        // onSave();
      })
      .catch((error) => {
        setApiError(true);
        console.error(error);
      })
      .finally(() => {
        setLoading(false);
      });
  }, [savedSearchName, user, t, validationError]);

  const onClose = useCallback(() => {
    setUpdated(false);
    setApiError(false);
    setLoading(false);
    setValidationError(undefined);
    setSavedSearchName("");
  }, []);

  return (
    <>
      <ModalToggleButton
        modalRef={modalRef}
        opener
        data-testid="open-save-search-modal-button"
        type="button"
        className="padding-1 hover:bg-base-lightest"
        unstyled
      >
        <USWDSIcon name="edit" key="edit-saved-search" />
        {editText}
      </ModalToggleButton>
      <Modal
        ref={modalRef}
        forceAction
        className="text-wrap"
        aria-labelledby={`${modalId}-heading`}
        aria-describedby={`${modalId}-description`}
        id={modalId}
        onKeyDown={(e) => {
          if (e.key === "Enter") handleSubmit();
        }}
      >
        {updated ? (
          <SuccessContent
            modalRef={modalRef}
            modalId={modalId}
            onClose={onClose}
          />
        ) : (
          <>
            <ModalHeading id={`${modalId}-heading`}>{t("title")}</ModalHeading>
            <div className="usa-prose">
              <p className="font-sans-2xs margin-y-4">
                {t.rich("description", {
                  strong: (chunks) => <strong>{chunks}</strong>,
                })}
              </p>
            </div>
            {loading ? (
              <Loading />
            ) : (
              <>
                {apiError && (
                  <SimplerAlert
                    alertClick={() => setApiError(false)}
                    buttonId="saveSearchApiError"
                    messageText={t("apiError")}
                    type="error"
                  />
                )}
                <SaveSearchInput
                  validationError={validationError}
                  updateSavedSearchName={setSavedSearchName}
                  id={savedSearchId}
                />
                <ModalFooter>
                  <Button
                    type={"button"}
                    onClick={handleSubmit}
                    data-testid="save-search-button"
                  >
                    {t("saveText")}
                  </Button>
                  <ModalToggleButton
                    modalRef={modalRef}
                    closer
                    unstyled
                    className="padding-105 text-center"
                    onClick={onClose}
                  >
                    {t("cancelText")}
                  </ModalToggleButton>
                </ModalFooter>
              </>
            )}
          </>
        )}
      </Modal>
    </>
  );
}
