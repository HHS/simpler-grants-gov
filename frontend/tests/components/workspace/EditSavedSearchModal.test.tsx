import { act, fireEvent, render, screen } from "@testing-library/react";
import { useTranslationsMock } from "src/utils/testing/intlMocks";

import { EditSavedSearchModal } from "src/components/workspace/EditSavedSearchModal";

const mockUseUser = jest.fn(() => ({
  user: {
    token: "faketoken",
  },
}));

const clientFetchMock = jest.fn();
const routerPush = jest.fn(() => Promise.resolve(true));

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(() => "/test") as jest.Mock<string>,
  useRouter: () => ({
    push: routerPush,
  }),
}));

jest.mock("src/services/auth/useUser", () => ({
  useUser: () => mockUseUser(),
}));

jest.mock("src/hooks/useClientFetch", () => ({
  useClientFetch: () => ({
    clientFetch: (...args: unknown[]) => clientFetchMock(...args) as unknown,
  }),
}));

jest.mock("next-intl", () => ({
  useTranslations: () => useTranslationsMock(),
}));

jest.useFakeTimers();

describe("EditSavedSearchModal", () => {
  afterEach(() => {
    clientFetchMock.mockReset();
    jest.clearAllTimers();
  });

  it("matches snapshot", async () => {
    const { container } = render(
      <EditSavedSearchModal
        savedSearchId="1"
        editText="edit"
        queryName="excellent query"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-edit-saved-search-modal-button-1",
    );
    act(() => toggle.click());

    expect(container).toMatchSnapshot();
  });
  it("displays a working modal toggle button", async () => {
    const { rerender } = render(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    expect(screen.queryByRole("dialog")).toHaveClass("is-hidden");

    const toggle = await screen.findByTestId(
      "open-edit-saved-search-modal-button-1",
    );
    act(() => toggle.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    expect(screen.getByRole("dialog")).not.toHaveClass("is-hidden");
  });
  it("modal can be closed as expected", async () => {
    const { rerender } = render(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-edit-saved-search-modal-button-1",
    );
    act(() => toggle.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const closeButton = await screen.findByText("cancelText");
    act(() => closeButton.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    expect(screen.queryByRole("dialog")).toHaveClass("is-hidden");
  });
  it("displays validation error if submitted without a name", async () => {
    const { rerender } = render(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-edit-saved-search-modal-button-1",
    );
    act(() => toggle.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const saveButton = await screen.findByTestId("edit-saved-search-button-1");
    act(() => saveButton.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const validationError = await screen.findByText("emptyNameError");

    expect(validationError).toBeInTheDocument();
  });
  it("displays an API error if API returns an error", async () => {
    clientFetchMock.mockRejectedValue(new Error());
    const { rerender } = render(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-edit-saved-search-modal-button-1",
    );
    act(() => toggle.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const saveButton = await screen.findByTestId("edit-saved-search-button-1");
    const input = await screen.findByTestId("textInput");
    fireEvent.change(input, { target: { value: "save search name" } });
    act(() => saveButton.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const error = await screen.findByText("apiError");

    expect(error).toBeInTheDocument();
  });

  it("displays a success message on successful save", async () => {
    clientFetchMock.mockResolvedValue({ id: "123" });
    const { rerender } = render(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const toggle = await screen.findByTestId(
      "open-edit-saved-search-modal-button-1",
    );
    act(() => toggle.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const saveButton = await screen.findByTestId("edit-saved-search-button-1");
    const input = await screen.findByTestId("textInput");
    fireEvent.change(input, { target: { value: "save search name" } });
    act(() => saveButton.click());

    rerender(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );

    const success = await screen.findByText("successTitle");

    expect(success).toBeInTheDocument();
  });
  it("defaults input to current name of query", () => {
    render(
      <EditSavedSearchModal
        queryName="excellent query"
        savedSearchId="1"
        editText="edit"
      />,
    );
    expect(screen.getByRole("textbox")).toHaveValue("excellent query");
  });
});
