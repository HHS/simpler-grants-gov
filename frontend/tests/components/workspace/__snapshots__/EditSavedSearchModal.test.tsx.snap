// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSavedSearchModal matches snapshot 1`] = `[Function]`;
